plugins/sharding/logical_count.rb

```diff
@@ -56,179 +41,115 @@ def cache_key(input)
-          if shard_key.nil?
-            message = "[logical_count] shard_key doesn't exist: " +
-                      "<#{shard.key_name}>"
+        def execute
+          have_shard = false
+          total = 0
+          each_shard_executor do |shard_executor|
+            have_shard = true
+            total += shard_executor.execute
+          end
+          unless have_shard
+            enumerator = @context.enumerator
+            message =
+              "[logical_count] no shard exists: " +
+              "logical_table: <#{enumerator.logical_table}>: " +
+              "shard_key: <#{enumerator.shard_key_name}>"
             raise InvalidArgument, message
           end
+          total
+        end
+      end
 
-          table_name = shard.table_name
-          range_index = nil
-          if @filter or @post_filter
-            log_use_range_index(false, table_name, "need filter",
-                                __LINE__, __method__)
-          elsif cover_type == :all
-            log_use_range_index(false, table_name, "covered",
-                                __LINE__, __method__)
-          else
-            index_info = shard_key.find_index(Operator::LESS)
-            range_index = index_info.index if index_info
-            if range_index
-              @referred_objects << range_index
-              log_use_range_index(true, table_name, "range index is available",
-                                  __LINE__, __method__)
-            else
-              log_use_range_index(false, table_name, "no range index",
-                                  __LINE__, __method__)
-            end
-          end
-          @contexts << ShardCountContext.new(shard, cover_type, range_index)
+      class ShardCountExecutor < StreamShardExecutor
+        include Loggable
+        include QueryLoggable
+
+        def initialize(context, shard, shard_range)
+          super("logical_count", context, shard, shard_range)
         end
 
-        def count
-          prepare_contexts
-          total = 0
-          @contexts.each do |context|
-            total += count_shard(context)
+        def execute
+          ensure_prepared
+          if @range_index
+            return count_n_records_in_range
           end
-          total
-        end
 
-        def close
-          @temporary_expressions.each(&:close)
-          @temporary_tables.each(&:close)
-          @referred_objects.each(&:unref)
-        end
+          ensure_filtered
+          return 0 if @filtered_result_sets.empty?
 
-        private
-        def log_use_range_index(use, table_name, reason, line, method)
-          message = "[logical_count]"
-          if use
-            message << "[range-index]"
-          else
-            message << "[select]"
+          total = 0
+          each_result_set do |result_set|
+            unless @post_filter.nil?
+              result_set = apply_post_filter(result_set)
+              @temporary_tables << result_set
+            end
```

Comment:
Can we move this to `StreamShardExecutor`?

Reply:
#1348 is merged